File: Implementations/C++/Algorithms/Sorting/mergeSort.cpp
#include <iostream>
#include <vector>
#include <algorithm>

typedef std::vector<int> vi;

void merge(vi& arr, const int left, const int mid, const int right)
{
    const int leftArrSize = mid - left + 1;
    vi leftArr(leftArrSize);

    const int rightArrSize = right - mid;
    vi rightArr(rightArrSize);

    // copy data from arr to the two temporary arrays
    for (int i = 0; i < leftArrSize; ++i) {
        leftArr[i] = arr[left + i];
    }
    for (int i = 0; i < rightArrSize; ++i) {
        rightArr[i] = arr[mid + 1 + i];
    }

    // create int variables to store the initial index of the two temporary arrays and the merged array
    int indexOfLeftArr{}, indexOfRightArr{}, indexOfMergedArr{left};

    // merge the two temporary arrays while comparing the elements from the two temporary arrays
    while (indexOfLeftArr < leftArrSize && indexOfRightArr < rightArrSize) {
        // if the element in the left array is smaller than the element in the right array, 
        // then copy the element from the left array to the merged array
        if (leftArr[indexOfLeftArr] <= rightArr[indexOfRightArr]) 
        {
            arr[indexOfMergedArr] = leftArr[indexOfLeftArr];
            indexOfLeftArr++;
        } 
        // if the element in the left array is larger than the element in the right array, 
        // then copy the element from the right array to the merged array
        else 
        {
            arr[indexOfMergedArr] = rightArr[indexOfRightArr];
            indexOfRightArr++;
        }
        indexOfMergedArr++;
    }

    // copy any remaining elements of the temporary arrays to the merged array
    while (indexOfLeftArr < leftArrSize) {
        arr[indexOfMergedArr] = leftArr[indexOfLeftArr];
        indexOfLeftArr++;
        indexOfMergedArr++;
    }
    while (indexOfRightArr < rightArrSize) {
        arr[indexOfMergedArr] = rightArr[indexOfRightArr];
        indexOfRightArr++;
        indexOfMergedArr++;
    }
}

/*
    * Merge Sort Details:
    * This is a sorting algorithm that works by dividing the array into two parts and sorting them recursively (divide and conquer).
    * The algorithm then merges the two sorted parts to form the final sorted array.
    * This algorithm is stable.

    * Best time complexity: O(n log n)
    * Worst time complexity: O(n log n)
    * Average time complexity: O(n log n)

    * Space complexity: O(n)

    * How it works:
    * 1. The algorithm starts by dividing the array into two parts
    * 2. Then it sorts the two parts recursively
    * 3. Then it merges the two sorted parts to form the final sorted array
*/
void mergeSort(vi& arr, const int start, const int end)
{
    if (start >= end) return; // if the array has LESS THAN 2 elements (x < 2), there is nothing to sort

    int mid = start + (end - start) / 2;
    mergeSort(arr, start, mid); // sort the left half
    mergeSort(arr, mid + 1, end); // sort the right half
    merge(arr, start, mid, end); // merge the two halves
}

int main()
{
    vi arr;
    std::cout << "Enter number of elements: ";
    int n; std::cin >> n;
    std::cout << "\n";

    std::cout << "Enter elements (with spaces as the delimiter):\n";
    for (int i=0; i < n; i++) {
        int x; std::cin >> x;
        arr.push_back(x);
    }

    std::cout << "\nSorted elements:\n";
    mergeSort(arr, 0, n - 1);
    for (const auto& i: arr) std::cout << i << " ";
    return 0;
}

File: NYP-DSA-Assignment/src/test_algorithms/merge_sort.py
"""
Not implemented for the assignment since this algorithm is taught in the module.
Hence, there was no extra marks for implementing this algorithm in the assignment.
"""

def merge_sort(arr:list[int], reverse:bool=False) -> None:
    """
    Sort the array using the merge sort algorithm.
    
    This is a sorting algorithm that works by dividing the array into two parts and 
    sorting them recursively (divide and conquer).
    The algorithm then merges the two sorted parts to form the final sorted array.
    
    Best time complexity: O(n log n)
    Worst time complexity: O(n log n)
    Average time complexity: O(n log n)
    
    Space complexity: O(n)
    
    Requires two arguments:
    - arr (list): The array to be sorted
    - reverse (bool): Whether the array is sorted in ascending or descending order. Defaults to False.
    """
    if (len(arr) > 1):
        mid = len(arr) // 2
        left = arr[:mid]
        right = arr[mid:]

        # Recursive call on each half
        merge_sort(left, reverse=reverse)
        merge_sort(right, reverse=reverse)

        # i and j iterators for traversing the two halves
        # k iterators for the main array
        i = j = k = 0

        while (i < len(left) and j < len(right)):
            if (not reverse and left[i] <= right[j]):
                arr[k] = left[i]
                # Move the iterator forward
                i += 1
            elif (reverse and left[i] >= right[j]):
                arr[k] = left[i]
                # Move the iterator forward
                i += 1
            else:
                arr[k] = right[j]
                j += 1

            # Move to the next slot
            k += 1

        # place all the remaining values from the left half
        for i in range(i, len(left)):
            arr[k] = left[i]
            k += 1

        # place all the remaining values from the right half
        for j in range(j, len(right)):
            arr[k] = right[j]
            k += 1